src/backend/utils/activity/memtrack.c

```diff
@@ -0,0 +1,165 @@
+	update_global_allocation(0, PG_ALLOC_INIT);
+
+	/* If we get a late request, send it to the long path. */
+	allocation_lower_bound = 0;
+	allocation_upper_bound = 0;
+}
+
+
+/*
+ * Update backend memory allocation for a new request.
+ *
+ * There are two versions of this function. This one, which updates
+ * global values in shared memory, and an optimized update_local_allocation()
+ * which only updates private values.
+ *
+ * This routine is the "slow path". We invoke it periodically to update
+ * global values and pgstat statistics.
+ */
+bool
+update_global_allocation(int64 size, pg_allocator_type type)
+{
+	int64		delta;
+	int64		dsm_delta;
+	uint64		dummy;
+
+	/* If we are still initializing, only update the private counters */
+	if (ProcGlobal == NULL || MyProcPid == 0)
+		return update_local_allocation(size, type);
+
+	/* Verify totals are not negative. Should never happen. */
+	Assert((int64) pg_atomic_read_u64(&ProcGlobal->total_memory_bytes) >= 0);
+	Assert((int64) pg_atomic_read_u64(&ProcGlobal->shared_memory_bytes) >= 0);
+
+	/* Calculate total bytes allocated or freed since last report */
+	delta = my_memory.allocated_bytes + size - reported_memory.allocated_bytes;
+
+	/*
+	 * If memory limits are set, and we are increasing our allocation, and we
+	 * not the postmaster...
+	 */
+	if (max_total_memory_bytes > 0 && delta > 0 && MyProcPid != PostmasterPid)
+	{
+		/* Update the global total memory counter subject to the upper limit. */
+		if (!pg_atomic_fetch_add_limit_u64(&ProcGlobal->total_memory_bytes, delta, max_total_memory_bytes, &dummy))
+			return false;
+	}
+
+	/*
+	 * Otherwise, update the global counter with no limit checking.
+	 */
+	else
+		pg_atomic_fetch_add_u64(&ProcGlobal->total_memory_bytes, delta);
+
+	/*
+	 * Update the private memory counters. This must happen after the limit is
+	 * checked.
+	 */
+	update_local_allocation(size, type);
+
+	/*
+	 * Update the global dsm memory counter to reflect changes since our last
+	 * report
+	 */
+	dsm_delta = my_memory.allocated_bytes_by_type[PG_ALLOC_DSM] -
+		reported_memory.allocated_bytes_by_type[PG_ALLOC_DSM];
+	pg_atomic_fetch_add_u64(&ProcGlobal->shared_memory_bytes, dsm_delta);
+
+	/* Update pgstat statistics if we are initialized as a backend process. */
+	if (MyBEEntry != NULL)
+	{
+		PGSTAT_BEGIN_WRITE_ACTIVITY(MyBEEntry);
+		MyBEEntry->st_memory = my_memory;
+		PGSTAT_END_WRITE_ACTIVITY(MyBEEntry);
+	}
+
+	/* Remember the values we just reported */
+	reported_memory = my_memory;
+
+	/* Update bounds so they bracket our new allocation size. */
+	allocation_upper_bound = my_memory.allocated_bytes + allocation_allowance_refill_qty;
+	allocation_lower_bound = my_memory.allocated_bytes - allocation_allowance_refill_qty;
+
+	/* Verify totals are not negative. Should never happen. */
+	Assert((int64) pg_atomic_read_u64(&ProcGlobal->total_memory_bytes) >= 0);
+	Assert((int64) pg_atomic_read_u64(&ProcGlobal->shared_memory_bytes) >= 0);
```

Comment:
This assert is failing when we're going through a crash-restart cycle due to a backend getting hit with a kill -9.  You can see this by doing a make check-world and/or just going into src/test/recovery and running 'make check' with TAP tests enabled.  Here's what the end of the PG server crash/restart log looks like, in case it's helpful:


> 2023-09-13 13:58:27.476 EDT [215504] LOG:  server process (PID 215533) was terminated by signal 9: Killed
> 2023-09-13 13:58:27.476 EDT [215504] DETAIL:  Failed process was running: INSERT INTO alive VALUES(in-progress-before-sigkill) RETURNING status;
> 2023-09-13 13:58:27.476 EDT [215504] LOG:  terminating any other active server processes
> 2023-09-13 13:58:27.478 EDT [215504] LOG:  all server processes terminated; reinitializing
> TRAP: failed Assert("(int64) pg_atomic_read_u64(&ProcGlobal->shared_memory_bytes) >= 0"), File: "src/backend/utils/activity/memtrack.c", Line: 162, PID: 215504
> bin/postgres(ExceptionalCondition+0x6e)[0x55ebf9f30726]
> bin/postgres(exit_tracked_memory+0x0)[0x55ebf9deee75]
> bin/postgres(dsm_impl_op+0x19f)[0x55ebf9da6924]
> bin/postgres(+0x4ddece)[0x55ebf9da3ece]
> bin/postgres(shmem_exit+0xf2)[0x55ebf9da6a18]
> bin/postgres(+0x467670)[0x55ebf9d2d670]
> bin/postgres(+0x468854)[0x55ebf9d2e854]
> bin/postgres(+0x469051)[0x55ebf9d2f051]
> bin/postgres(BackgroundWorkerInitializeConnection+0x0)[0x55ebf9d305f9]
> bin/postgres(main+0x228)[0x55ebf9c47216]
> /lib/x86_64-linux-gnu/libc.so.6(+0x29d90)[0x7f2a6fc29d90]
> /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80)[0x7f2a6fc29e40]
> bin/postgres(_start+0x25)[0x55ebf99a0445]